src/main/java/org/jabref/logic/pdf/search/indexing/IndexingTaskManager.java

```diff
@@ -0,0 +1,188 @@
+package org.jabref.logic.pdf.search.indexing;
+
+import java.util.List;
+
+import org.jabref.gui.util.BackgroundTask;
+import org.jabref.gui.util.TaskExecutor;
+import org.jabref.logic.importer.ParserResult;
+import org.jabref.logic.l10n.Localization;
+import org.jabref.model.database.BibDatabase;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.LinkedFile;
+
+import org.apache.lucene.store.Directory;
+
+/**
+ * Wrapper around {@link PdfIndexer} to execute all operations in the background.
+ */
+public class IndexingTaskManager extends BackgroundTask<Void> {
+
+    BackgroundTask<Void> runningTask;
+    TaskExecutor taskExecutor;
+
+    public IndexingTaskManager(TaskExecutor taskExecutor) {
+        this.taskExecutor = taskExecutor;
+        showToUser(true);
+        // the task itself is a nop, but it's progress property will be updated by the child-tasks it creates that actually interact with the inex
+        this.updateProgress(1, 1);
+        this.titleProperty().set(Localization.lang("Indexing pdf files"));
+        this.executeWith(taskExecutor);
+    }
+
+    @Override
+    protected Void call() throws Exception {
+        // update index to make sure it is up to date
+        this.updateProgress(1, 1);
+        return null;
+    }
+
+    public Directory getIndexDirectory() {
+        return this.getIndexDirectory();
+    }
+
+    private void enqueueTask(BackgroundTask<Void> task) {
+        // @todo is this any good? Is there a better way?
+        if (runningTask == null) {
+            task.executeWith(taskExecutor);
+            this.progressProperty().bind(task.progressProperty());
+        } else {
+            runningTask.onFinished(() -> {
+                task.executeWith(taskExecutor);
+                this.progressProperty().bind(task.progressProperty());
+            });
+        }
+        runningTask = task;
+        runningTask.onFinished(() -> {
+            this.progressProperty().unbind();
+            this.updateProgress(1, 1);
+            runningTask = null;
+        });
```

Comment:
Yes this is kind of hacky.
At the very least we need to change the name of the runningTask variable. It's not actually the running task, but the last task that was added.
Therefore, if more than one is in the queue, we make the new task run after the one that was last finished is done (onFinished).
I'll have a look at the Queue implementation, I'll just need to figure out how that can be connected to the BackgroundTask.

Reply:
Ah ok, it's already a linked list. that way.
A java Queue is just a FIFO list

Reply:
I misunderstood your original comment. I thought the Queue you were suggesting was already tailored for tasks, but its just the regular java.util.Queue. In that case it would not change much in terms of functionality, but it would be easier to read I think.

Reply:
Chose what's easier to read / clean code.

Reply:
I think an actual queue is more understandable. I love recursion, but in this case it's not necessary and an actual queue is better.
Implemented in 82b9d69